### ISSUE TITLE
Rule proposal: No default parameter options

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#208]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2482%20Funded-%2300A156.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/208)

I have seen this pattern of using default parameters with an object as default options values:

```js
const foo = (options = {a: false}) => {
	// …
};
```

This looks ok, but it's easy to miss that it only works with one option. If the user passes in an object, the whole default object will be replaced.

So here, the `b` option default value will be lost:

```js
const foo = (options = {a: false, b: true}) => {
	// …
};

foo({a: true});
```

I have seen this mistake a lot in the wild, so would be good to have a rule for it.

Example: https://github.com/lukechilds/byte-range/blob/3d154a9af0f827dfbb44c5acf32029cf1a549ccb/src/index.js#L3

The correct way to handle this is to use object spreading:

```js
const foo = options => {
	options = {
		a: false,
		b: true,
		...options
	};
};

foo({a: true});
```

The rule should fail on parameters named `options` or `opts` that have a default parameters value that is an object literal. The exception is an empty object, which should be allowed.

## Fail

```js
const foo = (options = {a: false}) => {
	// …
};
```

```js
const foo = (opts = {a: false, b: true}) => {
	// …
};
```

## Pass

```js
const foo = options => {
	options = {
		a: false,
		b: true,
		...options
	};
};
```

```js
const foo = (options = {}) => {
	// …
};
```

Would be nice to auto-fix this, but I don't think it's very easy, so it's not a requirement to resolve this issue.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $82.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($80.00)
- [<img src='https://avatars1.githubusercontent.com/u/40680108?v=4' alt='0maxxam0' width=24 height=24> 0maxxam0](https://issuehunt.io/u/0maxxam0) ($2.00)

### Submitted pull Requests
- [#633 Add `disallow-objects-as-default` rule](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/pull/633)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

**Comment by MrHen** [comment]
What about the following cases?

Variable default:

```
const fooDefaults = {
  a: false,
  b: true,
}
const foo = (options = fooDefaults) => {
	// …
};
foo({a: true});
```
----

Destructuring with defaults:

```
const foo = ({a = true, b = true}) => {
	// …
};
foo({a: true});
```

**Comment by sindresorhus** [comment]
> Variable default:

It's more complicated, yes, but would be nice to report on, although I doubt we'll be able to do a fixer for this case.

> Destructuring with defaults:

This should not be reported. It's a valid use of destructuring.

> This most works as expected. The only real gotcha is that the default value is only used to replace undefined -- it leaves null alone.

That's expected. It's how it works with non-deconstructing default parameters too.

**Comment by IssuehuntBot** [comment]
@issuehuntfest has funded $80.00 to this issue. [See it on IssueHunt](https://issuehunt.io/repos/55832243/issues/208)

**Comment by medusalix** [comment]
May I work on this issue?

**Comment by sindresorhus** [comment]
@medusalix Sure, go ahead :)

**Comment by futpib** [comment]
```js
const fooDefaults = {
  a: false,
  b: true,
}
const foo = (options = fooDefaults) => {}
```

> would be nice to report on

Should we really report this? With a bit of a stretch I can imagine confusing `(options = {a: false, b: true})` with `({ a = false, b = false } = {})`, but `(options = fooDefaults)` seems as clear as possible.

**Comment by sindresorhus** [comment]
> I imagine that it will report a lot of perfectly reasonable code

I'm confused. It's **not** perfectly reasonable. You above example also inhibits the problem that if one property is given, it does not use the default for the other properties. Or am I misunderstanding your point?

**Comment by futpib** [comment]
My point is that it's harder to misinterpret than the inline variant, and also that the rule will catch (in my estimation) reasonable code:
```js
const defaultContext = { /* imagine something complex */ };
function doSomething(context = defaultContext) {
  /* ... */
}
```
Unless the intentions is that the rule will only report when the argument name is "options", is that the intention here?

**Comment by futpib** [comment]
> The rule should fail on parameters named options or opts that have a default parameters value that is an object literal. The exception is an empty object, which should be allowed.
Sorry, ok, this makes sense now.

**Comment by medusalix** [comment]
@futpib Yeah, I added some test cases for your code example too 😃.

**Comment by futpib** [comment]
> The correct way to handle this is to use object spreading:
>
> ```js
> const foo = options => {
> 	options = {
> 		a: false,
> 		b: true,
> 		...options
> 	};
> };
>
> foo({a: true});
> ```

I don't think is is the correct way. It's better then the "fail" examples, but the correct way is to use default parameters or an utility function to properly merge defaults with the options.

```js
const foo = ({a = false, b = true} = {}) => {
  // ...
};
```

```js
const foo = (options) => {
  options = defaults(options, {
    a: false,
    b: true,
  });
};
```

This is better because it properly handles `undefined`, which is often used with the meaning of "default". That is, caller wishes that callee behave like no value has been passed (like if an object key or a positional argument is ommited). Examples of this are everywhere (fetch options, react's `defaultProps`, arguments in lodash).

I don't think we should fix to or even mention the object spread variant.

**Comment by medusalix** [comment]
@futpib I really like your solution compared to object spreading. Maybe we should fix the failing code to use default parameters instead? @sindresorhus What do you think?

**Comment by medusalix** [comment]
@sindresorhus What's your opinion on the proposed fix?

**Comment by IssuehuntBot** [comment]
@0maxxam0 has funded $2.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/55832243/issues/208) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/55832243/issues/208).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by sindresorhus** [comment]
> @sindresorhus What's your opinion on the proposed fix?

Sorry for totally dropping the ball on this. Yes, makes sense 👍

**Comment by medusalix** [comment]
@sindresorhus To be honest, I'm quite puzzled by the amount of different solutions proposed. Everyone seems to have their own way of handling default parameters and I'm not sure where to continue right now. See also the discussion right [here](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/29).

**Comment by fregante** [comment]
I see two separate rules:

- "disallow objects as defaults" unless they're empty objects (this issue)
- "prefer default parameters to `foo = foo || 2`", unless it's a (complex) object (https://github.com/sindresorhus/eslint-plugin-unicorn/issues/29)

### `prefer-default-parameters`

If you already use default parameters, this rule should never complain. This is more about moving obsolete "default handling" to ES6 default parameters, that's all (IMHO)


```js
function (a = 1) {} // ✅
function (a) {
	a = a || 1 // ❌
}
```

### `disallow-objects-as-default`

```js
function (a = 1) {} // ✅
function (a = {}) {} // ✅
function (a = {b: 1}) {} // ❌
```

Also, while this is probably a preferred fix for the last example...

```js
function ({b = 1} = {}) {} // ✅
```

... it requires `eslint` to replace all instances of `a.b` into `b`, which may not be possible. Sometimes you need the whole `a` object, so the safer _fix_ would probably be:


```js
function (a) {
	a = {
		b: 1,
		...a
	};
} // ✅
```

**Comment by sindresorhus** [comment]
@fregante That sounds good to me. However, the fixer you're suggesting might not be feasible as destructuring with default value and object-spread have slightly different behavior. Object-spread elements overwrite the defaults even if the value is undefined, destructuring default doesn't. Which makes the destructuring default a better alternative.

**Comment by fregante** [comment]
My last example should be fixed if `...a` was moved before `b`, I think that was my mistake. Are you referring to that one?

**Comment by sindresorhus** [comment]
I'm referring to that one, yes. The above example is definitely not correct. Now `b` will always override the input no matter what.

**Comment by fregante** [comment]
> Which makes the destructuring default a better alternative.

I suppose you’re talking about `function ({a=1, b=2})`. My suggestion (which is OP’s suggestion actually) preserves `arguments[0]` as an object, if you need it.

**Comment by sindresorhus** [comment]
Yes, but it still changes the behavior, so it cannot really auto-fix to that.
